Limit leader lease proposals to one outstanding per range

spencerkimball opened the pull request:
The intent of this change is to avoid endlessly proposing leader
lease commands on timeout expirations.

Remove the expiring context when getting leader leases. Instead if
there's an outstanding leader lease, all commands requiring it will
block until the proposed command succeeds or fails.

When a range is garbage collected, we return errors for all of its
pending Raft commands.

Because there's no longer a timeout for leader leases, the test rpc
sender in client_test.go had to be modified to more directly support
requests redirect to the existing leader replica instead of just
waiting for a timeout.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/5032)

<!-- Reviewable:end -->

spencerkimball committed:
Limit leader lease proposals to one outstanding per range

The intent of this change is to avoid endlessly proposing leader
lease commands on timeout expirations.

Remove the expiring context when getting leader leases. Instead if
there's an outstanding leader lease, all commands requiring it will
block until the proposed command succeeds or fails.

When a range is garbage collected, we return errors for all of its
pending Raft commands.

Because there's no longer a timeout for leader leases, the test rpc
sender in client_test.go had to be modified to more directly support
requests redirect to the existing leader replica instead of just
waiting for a timeout.

vivekmenezes commented:
Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

*[storage/replica.go, line 188 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032#-KCS2rLAZJR7JHMKF6mu:-KCS2rLBHUVIksUBYdd4:-853772944) ([raw file](https://github.com/cockroachdb/cockroach/blob/7bf8f44e2594376b636e84aeba323beeb766f4c1/storage/replica.go#L188)):*
might as well use
var replicaGCdError = roachpb.NewError(errors.New("this replica was GCd"))

---

*[storage/replica.go, line 271 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032#-KCS3OxYfp05YK3_h2CF:-KCS3OxZ1ov8gHlFvpBj:-911715209) ([raw file](https://github.com/cockroachdb/cockroach/blob/7bf8f44e2594376b636e84aeba323beeb766f4c1/storage/replica.go#L271)):*
Place this in a function so you can use: defer r.mu.Unlock()

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 5 files at r1.
Review status: 2 of 5 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[storage/replica.go, line 156 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032#-KCS54BgtUt_YWcOqQmm:-KCS54BhEuRCZQ39qLb1:-281819591) ([raw file](https://github.com/cockroachdb/cockroach/blob/7bf8f44e2594376b636e84aeba323beeb766f4c1/storage/replica.go#L156)):*
can you wrap these four members into an embedded struct? it's getting a bit crowded in here.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 5 of 5 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[storage/replica.go, line 188 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032#-KCS2rLAZJR7JHMKF6mu:-KCSCQiiuh1FlRagStYG:414839162) ([raw file](https://github.com/cockroachdb/cockroach/blob/7bf8f44e2594376b636e84aeba323beeb766f4c1/storage/replica.go#L188)):*
Better to keep it that way, not because what you're saying is a bad idea but because there's a higher risk of something not making a copy of the `*Error` but writing into it, in which case Weird Things™ would happen.

---

*[storage/replica.go, line 507 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032#-KCSDX0nliYWWRLPYJU_:-KCSDX0nliYWWRLPYJUa:-1959424598) ([raw file](https://github.com/cockroachdb/cockroach/blob/7bf8f44e2594376b636e84aeba323beeb766f4c1/storage/replica.go#L507)):*
what does the detail have to do with it? expected `if r.llErr != nil {`.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[storage/replica.go, line 188 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032#-KCS2rLAZJR7JHMKF6mu:-KCSEAV8I5n47RmXLAoT:1224371592) ([raw file](https://github.com/cockroachdb/cockroach/blob/7bf8f44e2594376b636e84aeba323beeb766f4c1/storage/replica.go#L188)):*
Also this error needs semantics - right now it'll just hit the client, but really it should be a `RangeMismatchError` or `RangeNotFoundError` to get the correct retry at `DistSender`.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
This is kind of at odds with the fix for #3299. Once that issue is fixed, leader lease requests won't stay in the pending state indefinitely so this won't really help (the channel recv in the goroutine will fire at about the time that the method would have returned and unlocked `llMu` in the old version). 

Instead, we need to either explicitly limit the number of entries that can be added to the raft log or make sure that the leader steps down when it can't make progress (etcd has a flag for this that I haven't turned on because I don't think it's compatible with coalesced heartbeats)

---

Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032#-:-KCSGuuAc7PA63rRqBIO:-1543639008)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: 4 of 5 files reviewed at latest revision, 4 unresolved discussions.

---

*[storage/replica.go, line 188 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032#-KCS2rLAZJR7JHMKF6mu:-KCSIHcyf4ipCrDX5P3X:-2120039491) ([raw file](https://github.com/cockroachdb/cockroach/blob/7bf8f44e2594376b636e84aeba323beeb766f4c1/storage/replica.go#L188)):*
Oh good point, @tschottdorf. A `RangeNotFoundError` is appropriate.

---

*[storage/replica.go, line 507 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032#-KCSDX0nliYWWRLPYJU_:-KCSQgrbphyVA2Av0hH6:-609445857) ([raw file](https://github.com/cockroachdb/cockroach/blob/7bf8f44e2594376b636e84aeba323beeb766f4c1/storage/replica.go#L507)):*
Changed.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
This change may obviate the need for a limit on raft log entries. What's going to put them there if a pending leader lease hasn't yet returned on a range? Well, perhaps if a server was constantly restarting and adding leader lease retries to the pending raft log. That seems like a pretty fringe possibility.

What's your timing look like for a fix for #3299? Because I really wanted to get something checked in so we can make forward progress on stability, which is gated by things inevitably falling down with this problem.

---

Review status: 4 of 5 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032#-:-KCSUG6L-HNvoOTReQpW:1309676365)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
In a partition, the last known leader will continue to accept client traffic and append commands to its local log. Or on second thought maybe it won't - once its lease expires it will stop, so the damage is limited to the amount of logs that can be generated in one lease interval.  So I guess this works too, and then we can close #3299 and won't need to worry about limiting the size of the raft log.

LGTM

---

Reviewed 4 of 5 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[storage/client_test.go, line 368 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032#-KCSdrMDsjDURG_hLUJW:-KCSdrMDsjDURG_hLUJX:249500771) ([raw file](https://github.com/cockroachdb/cockroach/blob/e11bd3a99ce8b5ba534e631087a2d6a54a91a913/storage/client_test.go#L368)):*
This comment is talking about stoppers, which are store-level and not range-level, so the original comment wa correct.

---

*[storage/replica.go, line 271 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032#-KCS3OxYfp05YK3_h2CF:-KCScoivvqESmzKPvK7H:2047739870) ([raw file](https://github.com/cockroachdb/cockroach/blob/7bf8f44e2594376b636e84aeba323beeb766f4c1/storage/replica.go#L271)):*
Needs to be a function, not just a curly-brace scope. 

---

*[storage/replica.go, line 441 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032#-KCSd1r9y4cTfCOHKZzd:-KCSd1r9y4cTfCOHKZze:-41472565) ([raw file](https://github.com/cockroachdb/cockroach/blob/e11bd3a99ce8b5ba534e631087a2d6a54a91a913/storage/replica.go#L441)):*
This log line and the one a few lines down should be V(1)

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032#-:-KCSWB0SS49t8-8ftuEc:1368976745)*
<!-- Sent from Reviewable.io -->

spencerkimball committed:
Alternate implementation which uses a channel for waiting on lease

This implementation has the big advantage of respecting the supplied
context's done channel and the stopper for upstream cancellations /
timeouts as well as process shutdown.

Had to upgrade the `multiTestContext` `rpcSend` method to allow timeouts
on replicas to trigger new RPCs to remaining replicas.

tamird commented:
Reviewed 2 of 5 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[storage/replica.go, line 507 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032#-KCSDX0nliYWWRLPYJU_:-KCTTPmP-FHZNONbPaUb:10599323) ([raw file](https://github.com/cockroachdb/cockroach/blob/7bf8f44e2594376b636e84aeba323beeb766f4c1/storage/replica.go#L507)):*
seems not changed

---

*[storage/replica.go, line 502 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032#-KCTTcJwClEkw8sVl7dc:-KCTTcJwClEkw8sVl7dd:-487273215) ([raw file](https://github.com/cockroachdb/cockroach/blob/e11bd3a99ce8b5ba534e631087a2d6a54a91a913/storage/replica.go#L502)):*
pErr is not nil, unlike specified in the message.
GetDetail is nil, unlike specified in the message.
The value of whichever thing is not nil should be included in the message

---

*[storage/replica.go, line 510 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032#-KCTTY6mUb1YlDKyevBJ:-KCTTY6nPqGO_9fDDXLJ:-1209250576) ([raw file](https://github.com/cockroachdb/cockroach/blob/e11bd3a99ce8b5ba534e631087a2d6a54a91a913/storage/replica.go#L510)):*
r.llMu.err is being inspected, not pErr.
r.llMu.err is not nil, unlike specified in the message.
GetDetail is nil, unlike specified in the message.
The value of whichever thing is not nil should be included in the message

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[storage/replica.go, line 271 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032#-KCS3OxYfp05YK3_h2CF:-KCWnLWn4zyW7b7iMgtE:1359363018) ([raw file](https://github.com/cockroachdb/cockroach/blob/7bf8f44e2594376b636e84aeba323beeb766f4c1/storage/replica.go#L271)):*
Yeah, good point; sometimes I forget this thanks to years of C++.

---

*[storage/replica.go, line 502 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032#-KCTTcJwClEkw8sVl7dc:-KCWnemtrZARbzQ2lQ8d:1961779687) ([raw file](https://github.com/cockroachdb/cockroach/blob/e11bd3a99ce8b5ba534e631087a2d6a54a91a913/storage/replica.go#L502)):*
This was just for testing...wanted to verify we never returned a non-nil pErr with a nil detail. It's gone now.

---

*[storage/replica.go, line 510 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032#-KCTTY6mUb1YlDKyevBJ:-KCWnlIeeYZDNQVTiB0n:-170854668) ([raw file](https://github.com/cockroachdb/cockroach/blob/e11bd3a99ce8b5ba534e631087a2d6a54a91a913/storage/replica.go#L510)):*
Again...just testing log statements. Sorry 'bout that.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
PTAL; I changed the mechanism from a cond var in order to respect upstream cancellations / timeouts and process shutdown machinery.

This will allow us to avoid senselessly stranding rpc goroutines which have long been cancelled by the caller.

tamird commented:
Reviewed 5 of 5 files at r3.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[storage/replica.go, line 511 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032#-KCXFLhWJuRmZJugszNl:-KCXFLhWJuRmZJugszNm:-288357900) ([raw file](https://github.com/cockroachdb/cockroach/blob/66f9cd711fdab244296b4262b26c955e545df7e7/storage/replica.go#L511)):*
this seems racy: this channel can be closed and replaced (along with r.mu.llErr) before the lock acquisition on the next line.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Reviewed 3 of 5 files at r3.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[storage/replica.go, line 421 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032#-KCXfAxjOM-NAGZsQEdr:-KCXfAxjOM-NAGZsQEds:1803369936) ([raw file](https://github.com/cockroachdb/cockroach/blob/66f9cd711fdab244296b4262b26c955e545df7e7/storage/replica.go#L421)):*
Whenever we clear `llChan`, we should close it to unblock any goroutines that may be waiting on it.

---

*[storage/replica.go, line 452 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032#-KCXfVMIqfvvYr-An2Hw:-KCXfVMIqfvvYr-An2Hx:468755634) ([raw file](https://github.com/cockroachdb/cockroach/blob/66f9cd711fdab244296b4262b26c955e545df7e7/storage/replica.go#L452)):*
This should probably be `RunAsyncTask`, not `RunWorker`.

---

*[storage/replica_test.go, line 498 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032#-KCXeLZDdL_akdwPQJ9F:-KCXeLZEgfR9ycEk_WJV:-1723222952) ([raw file](https://github.com/cockroachdb/cockroach/blob/7bf8f44e2594376b636e84aeba323beeb766f4c1/storage/replica_test.go#L498)):*
This test is important: the translation from `LeaseRejectedError` to `NotLeaderError` is one way that replicas removed from a group can discover this fact. It looks like all the necessary code is still in place, so why can't this test stay in?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
I rewrote the use of channels and mutexes around these leader leases. The new code is a lot cleaner, so PTAL.

Also had to add timeouts to the `multiTestContext` `rpcSend` method.

---

Review status: 1 of 7 files reviewed at latest revision, 9 unresolved discussions.

---

*[storage/replica.go, line 421 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032#-KCXfAxjOM-NAGZsQEdr:-KCbS2TIs7r9DRvGA2cc:-1165250762) ([raw file](https://github.com/cockroachdb/cockroach/blob/66f9cd711fdab244296b4262b26c955e545df7e7/storage/replica.go#L421)):*
Yeah, this code was f'd. It's been rewritten.

---

*[storage/replica_test.go, line 498 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032#-KCXeLZDdL_akdwPQJ9F:-KCY4i4ECGKf_2-wrCXT:-1461947908) ([raw file](https://github.com/cockroachdb/cockroach/blob/7bf8f44e2594376b636e84aeba323beeb766f4c1/storage/replica_test.go#L498)):*
Mistake. Added back.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032#-:-KCbRoFjw7D5slZKHWRL:623624182)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 6 of 6 files at r4.
Review status: all files reviewed at latest revision, 17 unresolved discussions, some commit checks failed.

---

*[storage/client_split_test.go, line 856 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032#-KCbW1KAo4bGPWfjKv8q:-KCbW1KBU2N55kWVgEYo:1122378450) ([raw file](https://github.com/cockroachdb/cockroach/blob/192816698933eb52cecb57602d7b122d4ecccdcc/storage/client_split_test.go#L856)):*
did you mean to leave this in?

---

*[storage/client_split_test.go, line 861 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032#-KCbW3eMfLkicNofDM6c:-KCbW3eMfLkicNofDM6d:-1677837463) ([raw file](https://github.com/cockroachdb/cockroach/blob/192816698933eb52cecb57602d7b122d4ecccdcc/storage/client_split_test.go#L861)):*
ditto.

---

*[storage/client_test.go, line 335 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032#-KCbXMSAMWQInTLjqBb4:-KCbXMSAMWQInTLjqBb5:-395364909) ([raw file](https://github.com/cockroachdb/cockroach/blob/192816698933eb52cecb57602d7b122d4ecccdcc/storage/client_test.go#L335)):*
nit: `for replicaIndex := range replicas {`

---

*[storage/client_test.go, line 337 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032#-KCbXg374etHI0gNJN_U:-KCbXg374etHI0gNJN_V:-585019476) ([raw file](https://github.com/cockroachdb/cockroach/blob/192816698933eb52cecb57602d7b122d4ecccdcc/storage/client_test.go#L337)):*
can you name this nodeID, make it type roachpb.NodeID, and add a comment where you cast it to `int` about the implicit node ID -> stopper index assumption?

The old code was more explicit in this regard.

---

*[storage/client_test.go, line 359 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032#-KCbYErrybAVvmDB3R3B:-KCbYErrybAVvmDB3R3C:-43033959) ([raw file](https://github.com/cockroachdb/cockroach/blob/192816698933eb52cecb57602d7b122d4ecccdcc/storage/client_test.go#L359)):*
if not nil, don't you want to know what the error is? doesn't seem harder to sent it on the channel and log it below.

---

*[storage/client_test.go, line 385 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032#-KCbYSJi8PXt9Mq3pX6P:-KCbYSJi8PXt9Mq3pX6Q:-1754846211) ([raw file](https://github.com/cockroachdb/cockroach/blob/192816698933eb52cecb57602d7b122d4ecccdcc/storage/client_test.go#L385)):*
ditto `for-range` is possible here

---

*[storage/replica.go, line 420 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032#-KCbWrwudASKiCG5ZzL1:-KCbWrwudASKiCG5ZzL2:2129737688) ([raw file](https://github.com/cockroachdb/cockroach/blob/192816698933eb52cecb57602d7b122d4ecccdcc/storage/replica.go#L420)):*
i think you can replace `llActive` with `len(r.mu.llChans) > 0`

---

*[storage/replica.go, line 483 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032#-KCbWXLJORCI_Sezzomr:-KCbWXLJORCI_Sezzoms:2085770915) ([raw file](https://github.com/cockroachdb/cockroach/blob/192816698933eb52cecb57602d7b122d4ecccdcc/storage/replica.go#L483)):*
change this to `r.mu.llChans = r.mu.llChans[:0]`

---

*[storage/replica.go, line 487 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032#-KCbWi5M8hF13UtAmWZA:-KCbWi5M8hF13UtAmWZB:1590546312) ([raw file](https://github.com/cockroachdb/cockroach/blob/192816698933eb52cecb57602d7b122d4ecccdcc/storage/replica.go#L487)):*
this should be an `append` if you adopt the change i suggested above

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032#-:-KCbYYrO5oZhcSTCrxVT:1417088721)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: all files reviewed at latest revision, 17 unresolved discussions, some commit checks failed.

---

*[storage/client_test.go, line 359 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032#-KCbYErrybAVvmDB3R3B:-KCbaSvQekrmROGWrgOe:-329644629) ([raw file](https://github.com/cockroachdb/cockroach/blob/192816698933eb52cecb57602d7b122d4ecccdcc/storage/client_test.go#L359)):*
I don't really want to log the error...it bubbles up; logging here will just end up being noisy.

---

*[storage/replica.go, line 420 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032#-KCbWrwudASKiCG5ZzL1:-KCbapVgCTG-NtMzDxoL:-1504763431) ([raw file](https://github.com/cockroachdb/cockroach/blob/192816698933eb52cecb57602d7b122d4ecccdcc/storage/replica.go#L420)):*
Good idea.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: 4 of 7 files reviewed at latest revision, 17 unresolved discussions.

---

*[storage/client_test.go, line 337 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032#-KCbXg374etHI0gNJN_U:-KCbZxWDEy_aTk4jsvp4:1207126603) ([raw file](https://github.com/cockroachdb/cockroach/blob/192816698933eb52cecb57602d7b122d4ecccdcc/storage/client_test.go#L337)):*
`nodeIndex` and `nodeID` are two different values (the ID of a node is always 1 more than its index), and in this case we use the index more than the ID.

---

*[storage/client_test.go, line 349 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032#-KCb_cqXjCwLj0sU8xRm:-KCb_cqXjCwLj0sU8xRn:-1405468779) ([raw file](https://github.com/cockroachdb/cockroach/blob/192816698933eb52cecb57602d7b122d4ecccdcc/storage/client_test.go#L349)):*
This looks thread-unsafe: the async task cannot rely on the lock acquired at the start of this method, since we do not guarantee that all of these tasks have stopped before returning from the outer method.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 3 of 3 files at r5.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[storage/client_test.go, line 359 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032#-KCbYErrybAVvmDB3R3B:-KCbc1SUQL1yQN9KrERX:-1295201536) ([raw file](https://github.com/cockroachdb/cockroach/blob/192816698933eb52cecb57602d7b122d4ecccdcc/storage/client_test.go#L359)):*
ok, but shouldn't this at least assert that this error is an expected one if it's not nil? it seems odd to discard whatever information is in there.

---

*[storage/replica.go, line 424 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032#-KCbceQfhhCRHN0xbTsV:-KCbceQg7GwYc55T1HQl:-1630413438) ([raw file](https://github.com/cockroachdb/cockroach/blob/e830251cf34dcfd58e6a584c01407f5dd74b2ab5/storage/replica.go#L424)):*
should this be a task? looks like it's gonna need to prevent shutdown to avoid flakiness.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[storage/client_test.go, line 337 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032#-KCbXg374etHI0gNJN_U:-KCbboH2G_cYeKEsfRKt:-1993495315) ([raw file](https://github.com/cockroachdb/cockroach/blob/192816698933eb52cecb57602d7b122d4ecccdcc/storage/client_test.go#L337)):*
I just took @tamird's suggestion. Seems fine.

---

*[storage/client_test.go, line 349 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032#-KCb_cqXjCwLj0sU8xRm:-KCbeUkQZ9mg7qjXOpep:-339431951) ([raw file](https://github.com/cockroachdb/cockroach/blob/192816698933eb52cecb57602d7b122d4ecccdcc/storage/client_test.go#L349)):*
OK, added some more complexity to ensure thread safety.

---

*[storage/client_test.go, line 359 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032#-KCbYErrybAVvmDB3R3B:-KCbd-893FoIGvaCn4Z0:321512557) ([raw file](https://github.com/cockroachdb/cockroach/blob/192816698933eb52cecb57602d7b122d4ecccdcc/storage/client_test.go#L359)):*
It's not discarded. It's returned on `sendChan`.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[storage/client_test.go, line 359 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032#-KCbYErrybAVvmDB3R3B:-KCbflfgD6-09u6HUlFc:910294729) ([raw file](https://github.com/cockroachdb/cockroach/blob/192816698933eb52cecb57602d7b122d4ecccdcc/storage/client_test.go#L359)):*
Ah, sorry about that, I misread.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Reviewed 3 of 6 files at r4, 2 of 3 files at r5, 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[storage/replica.go, line 424 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032#-KCbceQfhhCRHN0xbTsV:-KCbfLTf4464jh3So-u2:-110557208) ([raw file](https://github.com/cockroachdb/cockroach/blob/e830251cf34dcfd58e6a584c01407f5dd74b2ab5/storage/replica.go#L424)):*
A task isn't appropriate as there are cases where a raft group may not be available to grant the leader lease. Making this a task would mean those servers could never drain. Further, ensuring this finishes is not a guard against any sort of flakiness. Anything which depends on this should already be in a task.

A worker actually seems more appropriate, as we do want to shut down an extant request for leader lease when the stopper's `ShouldStop` receives. This is an internal goroutine which services external requests and should remain active as long as there are extant tasks in the system.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r7.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[storage/replica.go, line 424 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032#-KCbceQfhhCRHN0xbTsV:-KCcMrQfSTMyed3xgOCF:-996278278) ([raw file](https://github.com/cockroachdb/cockroach/blob/e830251cf34dcfd58e6a584c01407f5dd74b2ab5/storage/replica.go#L424)):*
OK, but you aren't selecting on `ShouldStop` - what is the worker buying you? also, starting a worker increments a waitgroup, which will be zero occasionally during shutdown, causing this call to panic (see the pattern in various grpc handlers where RunWorker is called inside of RunTask for this reason).

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[storage/replica.go, line 504 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032#-KCccVy3nnkoPdFv4TLK:-KCccVy3nnkoPdFv4TLL:1377105369) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd12c54949206255d3016c39c9fddb70dac404b0/storage/replica.go#L504)):*
if you're passing in the context, no point in also passing in the trace (because that's also in the context via `tracing.SpanFromContext()`

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[storage/replica.go, line 424 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032#-KCbceQfhhCRHN0xbTsV:-KCgul5H_uMMN4axIRwF:750766276) ([raw file](https://github.com/cockroachdb/cockroach/blob/e830251cf34dcfd58e6a584c01407f5dd74b2ab5/storage/replica.go#L424)):*
Workers are valuable even when they don't select on `ShouldStop` - they still block the "close" phase of the stopper. 

A Task seems to me like the right fit for the semantics of this function, and if using a task prevents a clean shutdown then it sounds like we have problems deeper in the stack. Of course, that may be tricky to fix and a worker may be an easier solution, but at the very least it's only safe to start workers at startup or in a task, so this needs to use the RunTask/RunWorker pattern pointed out elsewhere.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5032)*
<!-- Sent from Reviewable.io -->
